src/appleseed/renderer/modeling/object/curveobjectreader.cpp

```diff
@@ -75,369 +79,268 @@ namespace renderer
 
-    ifstream input;
-    input.open(filepath.c_str());
+        void push_vertex_opacity(const float o) override
+        {
+            return m_opacities.push_back(o);
+        }
 
-    if (!input.is_open())
-    {
-        RENDERER_LOG_ERROR("failed to open curve file %s.", filepath.c_str());
-        return object;
-    }
+        void push_vertex_color(const Color3f& c) override
+        {
+            return m_colors.push_back(GColor3(c));
+        }
 
-    Stopwatch<DefaultWallclockTimer> stopwatch;
-    stopwatch.start();
 
-    size_t curve1_count;
-    size_t curve3_count;
-    input >> curve1_count;
-    input >> curve3_count;
+      private:
+        const ParamArray    m_params;
+        const string        m_name;
+        CurveObject*        m_object;
+        size_t              m_split_count;
 
-    object->reserve_curves1(curve1_count);
-    object->reserve_curves3(curve3_count);
+        // Curve attributes and statistics
+        vector<GVector3>    m_vertices;
+        vector<GScalar>     m_widths;
+        vector<GScalar>     m_opacities;
+        vector<GColor3>     m_colors;
 
-    const size_t curve_count = curve1_count + curve3_count;
+        // Global statistics
+        size_t              m_total_vertex_count;
 
-    for (size_t c = 0; c < curve_count; ++c)
-    {
-        size_t control_point_count;
-        input >> control_point_count;
 
-        if (control_point_count != 2 && control_point_count != 4)
+        void reset_curve_variables()
         {
-            RENDERER_LOG_ERROR(
-                "while loading curve file %s: only linear curves (2 control points) or cubic curves (4 control points) are currently supported.",
-                filepath.c_str());
-            return object;
+            m_vertices.clear();
+            m_widths.clear();
+            m_opacities.clear();
+            m_colors.clear();
         }
 
-        if (control_point_count == 2)
-        {
-            GVector3 points[2];
-            GScalar widths[2];
 
-            for (size_t p = 0; p < control_point_count; ++p)
-            {
-                input >> points[p].x >> points[p].y >> points[p].z;
-                input >> widths[p];
+        void split_and_store(const Curve3Type& curve, const size_t split_count)
+        {
+            if (split_count > 0) {
+                Curve3Type child1, child2;
+                curve.split(child1, child2);
+                split_and_store(child1, split_count - 1);
+                split_and_store(child2, split_count - 1);
             }
-
-            // We never presplit degree-1 curves.
-            const Curve1Type curve(points, widths);
-            object->push_curve1(curve);
+            else m_object->push_curve3(curve);
         }
-        else
-        {
-            assert(control_point_count == 4);
 
-            GVector3 points[4];
-            GScalar widths[4];
+        void push_curve1()
+        {
+            for (int32 i = 0; i < m_vertices.size() - 1; ++i) {
```

Comment:
Type mismatch, won't build with VS. Use `size_t` for indentation. Fix everywhere in your code please.